### PR TITLE
[Kernel] Dynamic Per-Token Activation Quantization

### PR DESCRIPTION
**Pull request by dsikka**
# Summary

- Adds an additional `CompressedTensorsScheme` to support `w8a8` models with dynamic per-token activation quantization, `CompressedTensorsW8A8DynamicToken`. This scheme adds support for `w8a8` dynamic per-token models quantized through [sparseml](https://github.com/neuralmagic/sparseml) and saved through [compressed-tensors](https://github.com/neuralmagic/compressed-tensors)
- Expands `CompressedTensorsConfig` to use `QuantizationArgs`, `QuantizationStrategy` and `find_first_name_or_class_match` to help match the appropriate scheme to each layer.
-  Add `dynamic_int8_quant_kernel` CUDA kernel that performs int8 dynamic-per-token quantization
- Refactor utilities in `reduction_utils.cuh`

From Neural Magic, co-authored by @varun-sundar-rabindranath

### CODE REVIEW

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py [hunk 1]
```diff
@@ -15,6 +19,21 @@ def __init__(self, layer_quant_details: Dict[str, Any], ignore: List[str]):
         self.ignore = ignore
         self.layer_quant_details = layer_quant_details
 
+        llama_mapping = {
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py:22]: Can you explain what this is for and why this is needed?

File: vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_dynamictoken.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    def _shard_id_as_int(self, shard_id: Union[str, int]) -> int:
+        if isinstance(shard_id, int):
+            return shard_id
+
+        assert isinstance(shard_id, str)
+        qkv_idxs = {"q": 0, "k": 1, "v": 2}
+        assert shard_id in qkv_idxs
+        return qkv_idxs[shard_id]
+
+    def scales_shard_splitter(
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_dynamictoken.py:25]: Can these functions be moved to `CompressedTensorsScheme` ?

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py [hunk 3]
```diff
@@ -40,63 +59,79 @@ def get_quant_method(
...
+            module=layer,
+            targets=self.layer_quant_details.keys(),
+            check_contains=True)
+
         if layer_type_name is None:
             raise ValueError(f"Could not matching target for layer {layer}")
 
         layer_quant_details: Dict[str, Any] = self.layer_quant_details.get(
             layer_type_name, None)
+
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py:134]: nit - make PR cleaner

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py [hunk 4]
```diff
@@ -117,7 +152,9 @@ def create_weights(self, layer: torch.nn.Module,
                        **extra_weight_attrs):
         """
         Use the CompressedTensorsScheme associated with each layer to create 
-        the necessary parameters for the layer.
+        the necessary parameters for the layer. See LinearMethodBase for param
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py:155]: nit - make PR cleaner (dont touch this line)

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py [hunk 5]
```diff
@@ -139,7 +176,8 @@ def apply(self,
         """
         Use the output of create_weights and the CompressedTensorsScheme 
         associated with the layer to apply the forward pass with the 
-        layer input.
+        layer input.  See LinearMethodBase for param details
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py:179]: nit - make PR cleaner (dont touch this line)

File: vllm/_custom_ops.py [hunk 6]
```diff
@@ -264,21 +264,33 @@ def scaled_fp8_quant(
...
-    return q
+    if scale is not None:
+        # Static-per-tensor quantization.
+        vllm_ops.static_scaled_int8_quant(q, input, scale)
+        return q, scale
+
+    # Dynamic-per-token quantization.
+    input_scales = torch.empty((input.numel() // input.shape[-1], 1),
+                               dtype=torch.float32,
+                               device="cuda")
```
Comment (robertgshaw2-neuralmagic) [inline vllm/_custom_ops.py:291]: to keep consistent with fp8, maybe this should be `input.device`

File: vllm/_custom_ops.py [hunk 7]
```diff
@@ -264,21 +264,33 @@ def scaled_fp8_quant(
...
+            When not provided, we invoke dynamic-per-token quantization.
 
     Returns:
-        torch.Tensor: Output tensor in int8.
+      Tuple[Torch.Tensor, Torch.Tensor] : Output int8 tensor and scales.
     """
     q = torch.empty_like(input, dtype=torch.int8)
-    vllm_ops.static_scaled_int8_quant(q, input, scale)
-    return q
+    if scale is not None:
```
Comment (robertgshaw2-neuralmagic) [inline vllm/_custom_ops.py:283]: Can we make the names of the variables used internally in this function match the `scaled_fp8_quant` function?

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py [hunk 8]
```diff
@@ -40,63 +59,79 @@ def get_quant_method(
...
     def get_scheme(self, layer: torch.nn.Module) -> "CompressedTensorsScheme":
 
-        # TODO: update with matching function from `compressed_tensors`
-        layer_type_name = None
-        layer_name_class = type(layer).__name__.lower()
-        for target in self.layer_quant_details:
-            if target.lower() in layer_name_class:
-                layer_type_name = target
-                break
+        # TODO: update/map layer_name for llama models before
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py:121]: Can you explain this comment?

File: vllm/model_executor/layers/quantization/compressed_tensors/utils.py [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+    )
+    observer_kwargs: Dict[str, Any] = Field(
+        default_factory=dict,
+        description=
+        ("optional dict of kwargs to be passed directly to torch quantization "
+         "Observers constructor excluding quantization range or symmetry"),
+    )
+
+
+def find_first_name_or_class_match(
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/layers/quantization/compressed_tensors/utils.py:70]: Can you just add a brief comment on what these functions are for?

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py (hunk 4 above)
Comment (dsikka) [reply vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py:155]: The comment was expanded.

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py (hunk 5 above)
Comment (dsikka) [reply vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py:179]: The comment was expanded

File: vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_dynamictoken.py (hunk 2 above)
Comment (dsikka) [reply vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_dynamictoken.py:25]: Although they're the same, not guaranteed to be the same for other schemes that may be added so I think it makes sense to keep it in the scheme for now.

File: vllm/_custom_ops.py (hunk 6 above)
Comment (varun-sundar-rabindranath) [reply vllm/_custom_ops.py]: Fixed it 👍

File: vllm/_custom_ops.py (hunk 7 above)
Comment (varun-sundar-rabindranath) [reply vllm/_custom_ops.py:283]: renamed `q` to `output`. I believe the other variables are good as it is. please take a look. Thanks.

**Review by robertgshaw2-redhat — COMMENTED** [review]
LGTM.

File: csrc/quantization/compressed_tensors/int8_quant_kernels.cu [hunk 10]
```diff
@@ -38,6 +39,38 @@ __global__ void static_scaled_int8_quant_kernel(
...
 }
+
+template <typename scalar_t, typename scale_type>
+__global__ void dynamic_scaled_int8_quant_kernel(
+    const scalar_t* __restrict__ input, int8_t* __restrict__ out,
+    scale_type scale, const int hidden_size) {
+  const int tid = threadIdx.x;
+  const int token_idx = blockIdx.x;
+
+  float amax_val = 0.0f;
```
[the remark this answers is not in the document]
Comment (varun-sundar-rabindranath) [reply csrc/quantization/compressed_tensors/int8_quant_kernels.cu]: Yes. `amax` is confusing.

File: tests/kernels/test_int8_quant.py [hunk 11]
```diff
@@ -10,21 +10,52 @@
 SCALE = [0.1, 0.5, 0.8, 1.2, 2.1]
 
 
+@pytest.mark.parametrize("num_tokens", NUM_TOKENS)
+@pytest.mark.parametrize("hidden_size", HIDDEN_SIZES)
```
[the remark this answers is not in the document]
Comment (varun-sundar-rabindranath) [reply tests/kernels/test_int8_quant.py:15]: Added hidden-sizes 5137 and 8193

File: csrc/reduction_utils.cuh [hunk 12]
```diff
@@ -21,29 +21,47 @@
...
+}  // namespace detail
+
+template <typename T>
+using ReduceFnType = T (*)(T, T);
+
+// Helper function to return the next largest power of 2
+static constexpr int _nextPow2(unsigned int num) {
+  if (num <= 1) return num;
+  return 1 << (CHAR_BIT * sizeof(num) - __builtin_clz(num - 1));
+}
```
[the remark this answers is not in the document]
Comment (comaniac) [reply csrc/reduction_utils.cuh:46]: We definitely need another refactoring for `csrc/quantization`...but I don't have an out-of-box solution for this ATM.

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py [hunk 13]
```diff
@@ -47,51 +51,63 @@ def from_config(cls, config: Dict[str, Any]) -> "CompressedTensorsConfig":
...
+
+    def _get_schema(self, weight_quant: BaseModel,
+                    input_quant: BaseModel) -> "CompressedTensorsScheme":
+        if self._is_static_tensor_w8a8(weight_quant, input_quant):
+            return CompressedTensorsW8A8StaticTensor()
 
-        is_8_bits = weight_bit == input_bit == 8
-        is_tensor = weight_strategy == input_strategy == "tensor"
-        is_symmetric = weight_symmetric and input_symmetric
+        elif self._is_dynamic_token_w8a8(weight_quant, input_quant):
```
Comment (comaniac) [inline vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py:98]: nit: No need to use `elif` when `if` has a return statement.
```suggestion
        if self._is_dynamic_token_w8a8(weight_quant, input_quant):
```

**Review by comaniac — APPROVED** [review]
LGTM

**Review by robertgshaw2-redhat — APPROVED** [review]

**Pull request merged**